pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java

```diff
@@ -1048,8 +1051,29 @@ protected void handleSend(CommandSend send, ByteBuf headersAndPayload) {
 
         startSendOperation();
 
-        // Persist the message
-        producer.publishMessage(send.getProducerId(), send.getSequenceId(), headersAndPayload, send.getNumMessages());
+        if (isTxn) {
+            final long txnIdMostBits = send.getTxnidMostBits();
+            final long txnIdLeastBits = send.getTxnidLeastBits();
+            final long sequenceId = send.getSequenceId();
+            final long producerId = send.getProducerId();
+            TxnID txnID = new TxnID(txnIdMostBits, txnIdLeastBits);
+            headersAndPayload.retain();
```

Comment:
why we only `retain` the buffer only for transactional sends?

Reply:
In the transaction buffer, the reference counter of the ByteBuf will add. So I retain the buffer to avoid can't use it in the transaction buffer after the netty release the buffer. And I will release it manually after publishing messages to the ledger.

Reply:
And I found the normal send also `retain` the buffer when it adds into the ledger.

Reply:
My point is that these two branches might have same retain logic. But only one is explicitly called in the code here. We should make the behaviors consistent to avoid the confusion.